src/System.Net.Http/src/System/Net/Http/HttpContent.cs

```diff
@@ -309,14 +305,44 @@ public Task CopyToAsync(Stream stream, TransportContext context)
                 throw new ArgumentNullException("stream");
             }
 
-            TaskCompletionSource<object> tcs = new TaskCompletionSource<object>();
+            TaskCompletionSource<object> tcs = new TaskCompletionSource<object>(this);
             try
             {
                 Task task = null;
                 if (IsBuffered)
                 {
-                    byte[] data = this.GetDataBuffer(_bufferedContent);
-                    task = stream.WriteAsync(data, 0, (int)_bufferedContent.Length);
+                    Debug.Assert(_bufferedContent.Position == 0, "CopyToAsync assumes that _bufferedContent.Position == 0.");
+#if NETNative
+                    // Copy directly out of the MemoryStream buffer
+                    ArraySegment<byte> data;
+                    if (_bufferedContent.TryGetBuffer(out data))
+                        task = stream.WriteAsync(data.Array, data.Offset, data.Count);
+                    else
+                    {
+#endif
+                    task = _bufferedContent.CopyToAsync(stream).ContinueWithStandard(tcs, (copyTask, state) =>
+                    {
+                        // Reset the position of _bufferedContent
+                        var innerTcs = (TaskCompletionSource<object>)state;
+                        var innerThis = (HttpContent)innerTcs.Task.AsyncState;
+                        innerThis._bufferedContent.Position = 0;
+
+                        if (copyTask.IsFaulted)
+                        {
+                            innerTcs.TrySetException(GetStreamCopyException(copyTask.Exception.GetBaseException()));
+                        }
+                        else if (copyTask.IsCanceled)
+                        {
+                            innerTcs.TrySetCanceled();
+                        }
+                        else
+                        {
+                            innerTcs.TrySetResult(null);
```

Comment:
You're correct, `_bufferedContent` should be set to a `LimitMemoryStream` which is always [writable](https://github.com/jamesqo/corefx/blob/http-content/src/System.Net.Http/src/System/Net/Http/HttpContent.cs#L716). I've updated the PR to just use `TryGetBuffer` and do a debug assert for that.